src/beatree/mod.rs

```diff
@@ -308,26 +323,13 @@ impl Tree {
             new_meta,
         );
 
-        {
-            let mut inner = self.shared.lock().unwrap();
-            inner.bbn_index = bbn_index;
-            // TODO: watch out for the lock contention here. If we change 50k values, that might
-            // result in 50k bottom-level branch nodes being released. On top of that, there are
-            // also upper-level branch nodes. We are also adding a lot of nodes one by one.
-            //
-            // If you think about it, it's all not necessary, because we only need to access to
-            // the BNP freelist, which should be used exclusively by the sync thread. But note
-            // once crucial thing: the obsolete branches are not released until the BBN index is
-            // overwritten in the line above.
-            //
-            // Also, the sync needs to query the branch pages in the course of the sync, so that's
-            // another indication for splitting the read/write parts.
-            for id in obsolete_branches {
-                inner.branch_node_pool.release(id);
-            }
+        // Take the shared lock again to complete the update to the new shared state
+        let mut inner = self.shared.lock().unwrap();
+        inner.staging = inner.staging.clone().difference(staged_changeset);
```

Comment:
I didn't pay attention to this part during the initial review and I think this is wrong, unless I am missing something.

In the design doc we used primary and secondary staging: the idea was that while commits collect changes in the primary staging, the sync will operate on the secondary staging. However, this turned out to be only a specification device and we don't really need to name secondary staging just because it could be stored on the stack. Now since in code there is no primary/secondary separation in the spec it turned into just "staging"  to not confuse the implementers. That being said, fundamentially the idea stayed the same.

Hence, finding difference is just extra work.

Reply:
Oh god dammit, you are right 🤦 

That was the initial idea between the primary and secondary split, so that the changes could still be served during the syncing...